Add Gradle capability declarations to detect duplicate Guava artifacts

lmcrean opened the pull request:
fixes #6666

## Problem

Users can accidentally include duplicate Guava artifacts (guava-jdk5, guava-base, sisu-guava, etc.) alongside the main Guava library, causing classpath conflicts and runtime issues that are difficult to debug.

## Solution
Declare that Guava provides the capabilities of known duplicate artifacts in module.json, following the existing google-collections pattern. This enables Gradle to detect and report conflicts at build time.

## Changes
- Added capability declarations for 4 duplicate Guava artifacts in all 4 variant sections of module.json:
  - com.google.guava:guava-base
  - com.google.guava:guava-jdk5
  - org.sonatype.sisu:sisu-guava
  - org.hudsonci.lib.guava:guava

Note: The servicemix capability was removed after testing revealed it caused conflicts between Gradle variants.

## Testing
```bash
# Build and install locally with module metadata
./mvnw install -pl guava -DskipTests -q

# Verify module metadata contains capability declarations
grep -E "(guava-base|guava-jdk5|sisu-guava|hudsonci)" \
  ~/.m2/repository/com/google/guava/guava/999.0.0-HEAD-jre-SNAPSHOT/guava-999.0.0-HEAD-jre-SNAPSHOT.module
# Expected: 20 lines (3 artifacts × 4 = 12, hudsonci × 4 = 4, total = 16 capability lines + 4 group lines)

# For Gradle users - test conflict detection after release
# Note: This will only show conflicts once the changes are in a released version
# The SNAPSHOT version may not resolve correctly from mavenLocal()
mkdir test-guava-conflict && cd test-guava-conflict
cat > build.gradle << 'EOF'
plugins { id 'java' }
repositories { mavenCentral() }
dependencies {
    implementation 'com.google.guava:guava:NEXT_RELEASE_VERSION'
    implementation 'com.google.guava:guava-jdk5:17.0'
}
EOF
gradle dependencies --configuration compileClasspath
# Expected after release: Capability conflict error
# Current behavior (without these changes): Both dependencies resolve without conflict
```

## Breaking Changes
Builds that currently (incorrectly) include both Guava and duplicate artifacts will now fail with a capability conflict error. Users must resolve by excluding the duplicate artifact or using Gradle's capability resolution.

**Why this breaking change is necessary:**
- Prevents silent runtime failures (NoSuchMethodError, ClassNotFoundException)
- Having duplicate Guava classes leads to unpredictable classloading behavior
- Build-time failure is preferable to production runtime failure
- Follows the established pattern already used for google-collections
- Simple fix: exclude the duplicate or explicitly choose which one to use
- "Fail fast, fail loud, fail at build time - not in production"

lmcrean committed:
Add Guava dependencies to module.json

This update introduces several new Guava-related dependencies in the module.json file, including guava-base, guava-jdk5, sisu-guava, and others. These additions enhance the project's support for Guava libraries.

RELNOTES=n/a

cpovirk commented:
Thanks. I took the liberty of pulling some more items from #6666 in https://github.com/google/guava/pull/7991. (Because of the specifics of how our process works, we always submit an internal CL, which generates a new GitHub PR, but it still gets attributed to the original contributor.) If you don't want your name to show up on the additions there, let me know, and I can add them in a separate change.

lmcrean committed:
Remove Apache ServiceMix Guava bundle dependencies from module.json

cpovirk commented:
Thanks! Hmm, I wonder what's special about that one? Really, I wonder if that's the one that's behaving _as expected_: Maybe Gradle should be [forcing us to resolve the conflict](https://github.com/google/guava/pull/3683#:~:text=offers%20simple%20options%20to%20resolve%20the%20conflict) for _all_ the libraries? Maybe, to make that happen, we need to list the various conflicting libraries as [dependencies in the integration tests](https://github.com/google/guava/blob/c37d51edf0698129bb2eed5095277c392752cbb0/integration-tests/gradle/build.gradle.kts#L186) and probably as [`extraLegacyDependencies`](https://github.com/google/guava/blob/c37d51edf0698129bb2eed5095277c392752cbb0/integration-tests/gradle/build.gradle.kts#L28), too? And then it would force us to make the capability-resolution edit that you're suggesting?

But wait, maybe we're seeing the conflict even in the PR's current form?

```
Execution failed for task ':androidAndroidConstraintCompileClasspathJava:testClasspath'.
> Could not resolve all files for configuration ':androidAndroidConstraintCompileClasspathJava:compileClasspath'.
   > Could not resolve com.google.collections:google-collections:1.0.
     Required by:
         project :androidAndroidConstraintCompileClasspathJava
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'org.hudsonci.lib.guava:guava:999.0.0-HEAD-android-SNAPSHOT' also provided by [com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(jreApiElements), com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(androidApiElements)]
   > Could not resolve com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT.
     Required by:
         project :androidAndroidConstraintCompileClasspathJava
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'org.hudsonci.lib.guava:guava:999.0.0-HEAD-android-SNAPSHOT' also provided by [com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(jreApiElements), com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(androidApiElements)]
   > Could not resolve com.google.guava:guava.
     Required by:
         project :androidAndroidConstraintCompileClasspathJava
      > Module 'com.google.guava:guava' has been rejected:
           Cannot select module with conflict on capability 'org.hudsonci.lib.guava:guava:999.0.0-HEAD-android-SNAPSHOT' also provided by [com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(jreApiElements), com.google.guava:guava:999.0.0-HEAD-android-SNAPSHOT(androidApiElements)]
```

I wonder if the single conflict created by pulling in `google-collections` turns out to be enough to trigger the whole thing? If so, maybe we need _only_ the capability-resolution edit that you're suggesting?

lmcrean committed:
Add capability resolution for additional Guava modules in build.gradle.kts

This update introduces resolution strategies for several new Guava-related capabilities, including listenablefuture, guava-base, guava-jdk5, sisu-guava, and others. These changes enhance the project's dependency management for Guava libraries, following the recent additions in module.json.

RELNOTES=n/a

cpovirk commented:
I don't think, thanks. I just keep having things come up, but I do have this thread starred for when I finally think I have a spare moment.

assincrono reviewed (approved):
LGTM!
